backend/src/order.ts

```diff
@@ -354,162 +310,173 @@ export function getAllForShow(show_id: number): Promise<IAdminOrderListItem[]> {
-    throw 'Hash verification failed!';
+    const resp = await handler.verifySuccess(req);
+    log.info('Payment success verification succeeded', {order_id});
+
+    if ( await updatePaymentStatusToPaid(order_id) ) {
+      log.log('debug', 'Sending tickets', {order_id});
+      await sendTickets(order_id);
+    }
+
+    log.log('debug', 'Getting updated order', {order_id});
+    const order = await get(order_id);
+    return order;
+  } catch (err) {
+    log.error('Payment success verification failed', {error: err, order_id});
+    throw err;
   }
+}
 
-  log.info('Verification ok', {order_id: order_id});
+async function updatePaymentStatusToPaid(order_id: number): Promise<boolean> {
+  let conn: IConnection;
+  
+  try {
+    conn = await db.beginTransaction();
+  } catch (err) {
+    log.error('Payment status update to paid failed', {error: err, order_id});
+    throw err;
+  }
 
-  return db.query('select status from nk2_orders where id = :order_id', {order_id: order_id})
-  .then((rows) => {
-    if (rows[0].status === 'paid') {
+  try {
+    const res = await db.query('select status from nk2_orders where id = :order_id', {order_id}, conn);
+    if(res[0].status === 'paid') {
       log.info('Order was already paid', {order_id: order_id});
-      return null;
+      await db.rollback(conn);
+      return false;
     }
-    return db.query('update nk2_orders set status = "paid", payment_id = :payment_id where id = :order_id', { order_id: order_id, payment_id: params.PAID})
-    .then(() => sendTickets(order_id))
-    .catch((err) => {
-      log.error('Updating payment status failed', {error: err, order_id: order_id});
-      return Promise.reject('Updating payment status failed');
-    });
-  })
-  .then(() => {
-    return get(order_id);
-  });
 
+    log.log('debug', 'Updating order payment status', {order_id});
+    await db.query('update nk2_orders set status = "paid" where id = :order_id',{order_id}, conn);
+    await db.commit(conn);
+    return true;
+  } catch (err) {
+    log.error('Updating payment status failed', {error: err, order_id});
+    await db.rollback(conn);
+    throw err;
+  }
+}
+
+export async function paymentCancelled(order_id: number, req: Request): Promise<any> {
+  try {
+    log.log('debug', 'Cancelling payment', {order_id});
+    const handler = await getPaymentHandler(order_id);
+
+    log.log('debug', 'Handle payment cancel verification');
+    await handler.verifyCancel(req);
+
+    log.info('Payment cancel verification succeeded', {order_id});
+    return await deleteCancelledOrder(order_id);
+  } catch (err) {
+    log.error('Payment cancel verification failed', {error: err,order_id});
+    throw err;
+  }
+}
+
+async function deleteCancelledOrder(order_id: number): Promise<any> {
+  let conn: IConnection;
+  try {
+    conn = await db.beginTransaction()
+  } catch (err) {
+    log.error('Payment cancelling failed', {error: err, order_id});
+  }
+
+  try {
```

Comment:
I'm not seeing the need to split this into two try-catches? Actually this is wrong since if creating the conn fails we'd still keep trying the rest.

If the separation is because of different stuff in the catch-block, maybe use just if(conn) await db.rollback(conn)?

Also the catch should probably rethrow the error

Reply:
Fixed on all cases.